man/man1/jpeg-compare.1
.TH "jpeg-compare" 1 2.6.4 "14 Feb 2023" "User manual"

.SH NAME
jpeg-compare

.SH DESCRIPTION
Compare two JPEG photos to judge how similar they are.
The fast comparison method returns an integer from 0 to 99, where 0 is identical.
PSNR, SSIM, and MS-SSIM return floats but require images to be the same dimensions.

.SH SYNOPSIS
jpeg-compare [options] image1.jpg image2.jpg

.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
output program help
.TP
\fB\-m\fR, \fB\-\-method\fR [arg]
set comparison method to one of 'fast', 'mpe', 'psnr', 'mse', 'msef', 'cor', 'ssim', 'ms-ssim', 'vifp1', 'smallfry', 'shbad', 'nhw', 'ssimfry', 'ssimshb', 'sum' [fast]
.TP
\fB\-n\fR, \fB\-\-norm\fR
UM scale metric
.TP
\fB\-r\fR, \fB\-\-ppm\fR
parse first input as PPM instead of JPEG
.TP
\fB\-s\fR, \fB\-\-size\fR [arg]
set fast comparison image hash size
.TP
\fB\-T\fR, \fB\-\-input-filetype\fR [arg]
set first input file type to one of 'auto', 'jpeg', 'ppm' [auto]
.TP
\fB\-U\fR, \fB\-\-second-filetype\fR [arg]
set second input file type to one of 'auto', 'jpeg', 'ppm' [auto]
.TP
\fB\-V\fR, \fB\-\-version\fR
output program version
.TP
\fB\-\-short\fR
do not prefix output with the name of the used method

.SH EXAMPLES
Do a fast compare of two images:
.PP
.I
jpeg-compare image1.jpg image2.jpg
.PP
Calculate PSNR:
.PP
.I
jpeg-compare --method psnr image1.jpg image2.jpg
.PP
Calculate SSIM:
.PP
.I
jpeg-compare --method ssim image1.jpg image2.jpg

.SH NOTES
"Universal Scale" of metrics (UM):
  0.0
  ... (DIRTY) ...
  0.5
  ... (LOW) ...
  0.75
  ... (MEDIUM) ...
  0.875
  ... (SUBHIGH) ...
  0.9375
  ... (HIGH) ...
  0.96875
  ... (VERYHIGH) ...
  1.0
.PP
Trends:

  UM = 0.29 * (sqrt(sqrt(255.0 / MPE)) - 1.0)
  UM = 0.557 * (sqrt(PNSR) - 5.0)
  UM = 0.5 * (sqrt(sqrt(1.0 / MSEF)) - 1.0)
  UM = 1.0 * cor_sigma(cor_sigma(COR))
  UM = 1.57 * cor_sigma(cor_sigma(cor_sigma(SSIM)))
  UM = 1.59 * cor_sigma(cor_sigma(MS_SSIM))
  UM = 1.10 * cor_sigma(cor_sigma(VIFP1))
  UM = 3.0 * (SMALLFRY * 0.01 - 0.8)
  UM = 1.46 * cor_sigma(SHARPENBAD)
  UM = 0.342 * (sqrt(sqrt(1.0 / NHW)) - 1.0)
    cor_sigma(M) = 1.0 - sqrt(1.0 - M * M)

.SH COPYRIGHT
 JPEG-Archive is copyright © 2015 Daniel G. Taylor
 Image Quality Assessment (IQA) is copyright 2011, Tom Distler (http://tdistler.com)
 SmallFry is copyright 2014, Derek Buitenhuis (https://github.com/dwbuiten)
 All rights reserved.

.SH "SEE ALSO"
 jpeg-hash,
 jpeg-recompress,
 jpeg-zfpoint,
 webp-compress,
 cjpeg


CHANGELOG
JPEG Recompress
https://github.com/ImageProcessing-ElectronicPublications/jpeg-recompress

2.6.4 "um"

    Research: correct UM model: all metrics.
    2023, zvezdochiot

2.6.3 "um"

    Research: correct UM model.
    2023, zvezdochiot

2.6.2 "sharpenbad"

    reimplementation SHARPENBAD:
      metric_sharpenbad

2.6.1 "msef"

    Added MSE and MSEF metric.
    Research: correct trends UM (44 images)

2.6.0 "sum"

    Added VIFP 1 layer in SUM metric.

2.5.9 "vifp1"

    Added VIFP 1 layer metric.

2.5.8 "neat"

    Neat adds up all values.

2.5.7 "bisection"

    Fix bisection

2.5.6 "subsample"

    Fix IPEP/#1: ignores subsampling option

2.5.5 "sigma"

    Move `sigma`-function from `libsmallfry` to `jmetrics.a`
    Research: correct trends UM (32 images)
    Remove metrics `cor` and `corsh`
    2022, zvezdochiot

2.5.4 "near"

    Research: correct trends UM (24 images)
    2022, zvezdochiot

2.5.3 "approx"

    Research: correct trends UM (16 images)
    2022, zvezdochiot

2.5.2 "UM"

    Research: correct trends UM (8 images)
    2022, zvezdochiot

2.5.1 "saved"

    Fix overflow saved.
    Correct quality scale (next = (1.0 + prev) / 2)
    Replacing multiplications with counters.
    2022, zvezdochiot

2.5.0 "webp"

    Use libwebp for WEBP format
    2022, zvezdochiot

2.4.0 "nhw"

    Update for libsmallfry 0.1.0
    metric_nhw
    2022, zvezdochiot

2.3.1 "weight"

    weight average for SUM metrics
    2021, zvezdochiot

2.3.0 "return"

    return from mozjpeg to jpeg8
    2021, zvezdochiot

2.2.1 "linearize"

    jpeg-recompress: linearize metric
    2018, zvezdochiot

2.2 "zfpoint"

    Find ZF point quality for JPEG
    Update for libsmallfry 0.0.3
    2018, zvezdochiot

2.1.3 "sum"

    jpeg-recompress: summary metric and linearize ssim
    2018, zvezdochiot

2.1.2

    Update for libsmallfry 0.0.2
    sharpenbad_metric
    2018, zvezdochiot

2.1.1

    fork jpeg-archive 2.1.1 to jpeg-recompress for compatible libsmallfry.
    Daniel G. Taylor & zvezdochiot


src/jpeg-compare.c
/*
    Compare two JPEG images. Several methods are available. PSNR, SSIM,
    and MS_SSIM require the images to be exactly the same size, while
    FAST works on any sized image.

    FAST compares two images and returns the difference on a
    scale of 0 - 99, where 0 would mean the images are identical. The
    comparison should be immune to exposure, saturation, white balance,
    scaling, minor crops and similar modifications.

    If the difference is 10 or less than the images are very likely
    different versions of the same image (e.g. a thumbnail or black
    and white edit) or just slightly different images. It is possible
    to get false positives, in which case a slower PSNR or SSIM
    comparison will help.
*/

#include <getopt.h>
#include "jmetrics.h"

void usage(char *progname)
{
    printf("usage: %s [options] image1.jpg image2.jpg\n\n", progname);
    printf("options:\n\n");
    printf("  -h, --help                   output program help\n");
    printf("  -m, --method [arg]           set comparison method to one of:\n");
    printf("                               'fast', 'mpe', 'psnr', 'mse', 'msef', 'cor', 'ssim', 'ms-ssim', 'vifp1',\n");
    printf("                               'smallfry', 'shbad', 'nhw', 'ssimfry', 'ssimshb', 'sum' [fast]\n");
    printf("  -n, --norm                   UM scale metric\n");
    printf("  -r, --ppm                    parse first input as PPM instead of JPEG\n");
    printf("  -s, --size [arg]             set fast comparison image hash size\n");
    printf("  -T, --input-filetype [arg]   set first input file type to one of 'auto', 'jpeg', 'ppm' [auto]\n");
    printf("  -U, --second-filetype [arg]  set second input file type to one of 'auto', 'jpeg', 'ppm' [auto]\n");
    printf("  -V, --version                output program version\n");
    printf("      --short                  do not prefix output with the name of the used method\n");
}

int main (int argc, char **argv)
{
    int method = FAST;
    int umscale = 0;

    int printPrefix = 1;

    // Hash size when method is FAST
    int size = 16;

    unsigned char *imageBuf1, *imageBuf2;
    long bufSize1, bufSize2;
    char *fileName1, *fileName2;

    // Use PPM input?
    enum filetype inputFiletype1 = FILETYPE_AUTO;
    enum filetype inputFiletype2 = FILETYPE_AUTO;

    const char *optstring = "VhS:m:nrT:U:";
    static const struct option opts[] =
    {
        { "version", no_argument, 0, 'V' },
        { "help", no_argument, 0, 'h' },
        { "size", required_argument, 0, 'S' },
        { "method", required_argument, 0, 'm' },
        { "norm", no_argument, 0, 'n' },
        { "ppm", no_argument, 0, 'r' },
        { "input-filetype", required_argument, 0, 'T' },
        { "second-filetype", required_argument, 0, 'U' },
        { "short", no_argument, 0, OPT_SHORT },
        { 0, 0, 0, 0 }
    };
    int opt, longind = 0;

    char *progname = "jpeg-compare";

    while ((opt = getopt_long(argc, argv, optstring, opts, &longind)) != -1)
    {
        switch (opt)
        {
        case 'V':
            version();
            return 0;
        case 'h':
            usage(progname);
            return 0;
        case 's':
            size = atoi(optarg);
            break;
        case 'm':
            method = parseMethod(optarg);
            break;
        case 'n':
            umscale = 1;
            break;
        case 'r':
            if (inputFiletype1 != FILETYPE_AUTO)
            {
                error("multiple file types specified for input file 1");
                return 1;
            }
            inputFiletype1 = FILETYPE_PPM;
            break;
        case 'T':
            if (inputFiletype1 != FILETYPE_AUTO)
            {
                error("multiple file types specified for input file 1");
                return 1;
            }
            inputFiletype1 = parseInputFiletype(optarg);
            break;
        case 'U':
            if (inputFiletype2 != FILETYPE_AUTO)
            {
                error("multiple file types specified for input file 2");
                return 1;
            }
            inputFiletype2 = parseInputFiletype(optarg);
            break;
        case OPT_SHORT:
            printPrefix = 0;
            break;
        };
    }

    if (argc - optind != 2)
    {
        usage(progname);
        return 255;
    }

    // Read the images
    fileName1 = argv[optind];
    fileName2 = argv[optind + 1];

    bufSize1 = readFile(fileName1, (void **)&imageBuf1);
    if (!bufSize1)
    {
        error("failed to read file: %s", fileName1);
        return 1;
    }

    bufSize2 = readFile(fileName2, (void **)&imageBuf2);
    if (!bufSize2)
    {
        error("failed to read file: %s", fileName2);
        return 1;
    }

    /* Detect input file types. */
    if (inputFiletype1 == FILETYPE_AUTO)
        inputFiletype1 = detectFiletypeFromBuffer(imageBuf1, bufSize1);
    if (inputFiletype2 == FILETYPE_AUTO)
        inputFiletype2 = detectFiletypeFromBuffer(imageBuf2, bufSize2);

    // Calculate and print output
    switch (method)
    {
    case FAST:
        if (inputFiletype1 != FILETYPE_JPEG || inputFiletype2 != FILETYPE_JPEG)
        {
            error("fast comparison only works with JPEG files!");
            return 255;
        }
        return compareFastFromBuffer(imageBuf1, bufSize1, imageBuf2, bufSize2, printPrefix, size);
    case MPE:
    case PSNR:
    case MSE:
    case MSEF:
    case SSIM:
    case MS_SSIM:
    case VIFP1:
    case SMALLFRY:
    case SHARPENBAD:
    case COR:
    case NHW:
    case SSIMFRY:
    case SSIMSHBAD:
    case SUMMET:
        return compareFromBuffer(method, imageBuf1, bufSize1, imageBuf2, bufSize2, printPrefix, umscale, inputFiletype1, inputFiletype2);
    default:
        error("unknown comparison method!");
        return 255;
    }

    // Cleanup resources
    free(imageBuf1);
    free(imageBuf2);
}
